### DISCUSSION TITLE
How to use sourcemap with remix dev ?

### DISCUSSION THREAD

**Discussion by Alex-Levacher** [remix-run/remix#8168] [Q&A]
Hello community,

I am currently facing an issue where I cannot find a solution. It appears that `remix dev` is not utilizing the sourcemaps.

When I simulate an error in `routes/_index.tsx` and then run `npm run dev`, it throws an error, as expected. Where the problem arises is that the error is not pointed out using source maps to facilitate the debugging process.

Here are the steps I followed:

1. I created a new project using `npx create-remix@latest`.
2. I simulated a syntax error in the `routes/_index.tsx` file.
3. I run the development script `npm run dev`.

This is when the error throws, but without pointing out the specific source of the error via a source map. It shows the error in the compiled file.

Is there something that I am missing here? Is there an additional step required to enable the use of source maps in this context?

Thank you so much for your time and assistance!
<img width="467" alt="Capture d’écran 2023-11-29 à 15 02 15" src="https://github.com/remix-run/remix/assets/8192049/7c32af17-7982-47bf-b68e-35076f536149">

<img width="1319" alt="Capture d’écran 2023-11-29 à 14 52 28" src="https://github.com/remix-run/remix/assets/8192049/2d9f41ac-de2c-4695-a8ea-8ea02cb455fa">

**Comment by kiliman** [comment]
Looks like sourcemaps broke in v2+. I know they changed it where you had to install `source-map-support` but that doesn't appear to be working even with Remix App Server. https://remix.run/docs/en/main/start/v2#source-map-support

Interestingly, debugging works in that I can put a breakpoint and it will stop at that line.

Exact same code:
```ts
import { Form } from "@remix-run/react";

export async function action() {
  throw new Error("Oops!");
}

export default function Index() {
  return (
    <div>
      <h1>Welcome to Remix</h1>
      <Form method="post">
        <button>Submit</button>
      </Form>
    </div>
  );
}
```

### Remix v1.19.3
<img width="756" alt="image" src="https://github.com/remix-run/remix/assets/47168/78b56e9c-6cc0-4e72-9c04-ef8dc6181713">

### Remix v2.0.0
<img width="941" alt="image" src="https://github.com/remix-run/remix/assets/47168/de771ecc-c81e-47de-8e37-2db9d1ecd53b">

### Breakpoints still work in v2.0+
<img width="540" alt="image" src="https://github.com/remix-run/remix/assets/47168/effdf079-7290-41a2-83f9-0423b65c9271">

**Comment by kiliman** [comment]
Ok, I think I figured out the issue.

I switched from using Remix App Server to Express and verified `source-map-support` was installed and still had the same issue.

I verified that `build/index.js` had the `//# sourceMappingURL=index.js.map` at the end of the file, and it did. And the map file looked fine.

Then I was thinking, perhaps it's not finding the map file because of the `?t=###` appended to the end of the `index.js` file. This is added when reimporting the build after a change.

Sure enough, as soon as I removed the timestamp, it worked.

@pcattori Not sure if this is a known issue or not.

```diff
// use a timestamp query parameter to bust the import cache
-   return import(BUILD_URL + "?t=" + stat.mtimeMs);
+   return import(BUILD_URL);
}
```

<img width="806" alt="image" src="https://github.com/remix-run/remix/assets/47168/0c32f50e-0b31-40d6-9e0e-4f90c03c0822">

It turns out that `install-source-maps` has a hook for retrieving the source map by source file.

```ts
sourceMapSupport.install({
  retrieveSourceMap: function (source) {
    // get source file without the `file://` prefix or `?t=...` suffix
    let match = source.match(/^file:\/\/(.*)\?t=[.\d]+$/);
    if (match) {
      return {
        url: source,
        map: fs.readFileSync(`${match[1]}.map`, "utf8"),
      };
    }
    return null;
  },
});
```

And now it works correctly with or without the `?t=` suffix.
```
GET /?index 200 - - 9.547 ms
retrieveSourceMap /Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/server-runtime/dist/server.js
retrieveSourceMap /Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/router/dist/router.cjs.js
retrieveSourceMap /Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/server-runtime/dist/routes.js
retrieveSourceMap /Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/server-runtime/dist/data.js
retrieveSourceMap file:///Users/michael/Projects/scratch/remix-2.0.0/build/index.js?t=1701291717719.318
Error: Oops!
    at action (file:///Users/michael/Projects/scratch/remix-2.0.0/app/routes/_index.tsx:4:9)
    at Object.callRouteActionRR (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/server-runtime/dist/data.js:35:22)
    at handler (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/server-runtime/dist/routes.js:61:50)
    at runHandler (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/router/router.ts:3621:7)
    at callLoaderOrAction (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/router/router.ts:3681:22)
    at submit (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/router/router.ts:2826:22)
    at queryImpl (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/router/router.ts:2761:28)
    at Object.queryRoute (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/router/router.ts:2711:24)
    at handleDataRequestRR (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/server-runtime/dist/server.js:77:40)
    at requestHandler (/Users/michael/Projects/scratch/remix-2.0.0/node_modules/@remix-run/server-runtime/dist/server.js:51:24)
```

**Reply by brettdh** [reply]
Is this hook going to be required for any server adapter that wants to use `source-map-support`? I'm wondering if:
- there should be some dedicated docs now for ensuring that source maps work, rather than just a note in the v2 upgrade guide
- source map support should be the responsibility of the adapter or the app that uses the adapter (and whether Remix docs should be opinionated about this)

**Reply by kiliman** [reply]
I wonder why source map handling is a manual concern instead of something automatic. Hopefully, @pcattori may be able to answer that.

Also, will this still be an issue as we proceed with Vite?

**Comment by lime517** [comment]
Thrilled to see other eyes on this issue, and thank you @kiliman for digging in and finding the cause!

**Reply by kiliman** [reply]
I've added a PR for the Remix App Server. I'll also create a PR for the Express template.

https://github.com/remix-run/remix/pull/8174

**Reply by kiliman** [reply]
Added PR for Express template

https://github.com/remix-run/remix/pull/8176

**Reply by Alex-Levacher** [reply]
Thank you, @kiliman, for suggesting those two fixes. I'm unsure about something: Does your #8174 fix also apply to remix dev? Does remix dev use remix-serve?

**Reply by kiliman** [reply]
@Alex-Levacher Yes, #8174 will include the fix automatically for `remix-serve` aka Remix App Server.

**Reply by Alex-Levacher** [reply]
Awesome !

**Reply by cliffordfajardo** [reply]
is source-map-support needed for remix express on vite?
i noticed the new remix vite express template no longer has the sourcemap code fix that @kiliman made here
https://github.com/remix-run/remix/pull/8174

**Reply by fatso83** [reply]
We did not have source maps, but after adding it to our Express on Vite setup it works.
